PHOENIX-6888 Fixing TTL and Max Lookback Issues for Phoenix Tables

kadirozde opened the pull request.

stoty commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:
```diff
@@ -426,58 +421,39 @@ RegionScanner getWrappedScanner(final ObserverContext<RegionCoprocessorEnvironme
     }
 
+    @Override
+    public InternalScanner preCompact(ObserverContext<RegionCoprocessorEnvironment> c, Store store,
```
This disables the logic for the old indexing in UngroupedAggregateRegionObserver#preCompactScannerOpen .

We should add the capability to disable delete processing fully in StoreCompactionScanner (or subclass it), and 
move that logic to UngroupedAggregateRegionObserver#preCompact() to set  up the Scanner to never remove Deleted cells.

kadirozde replied:
Thank you for pointing out this! I will make the necessary changes.

kadirozde replied:
Done

apurtell commented on phoenix-core/src/main/java/org/apache/phoenix/filter/PagedFilter.java:
```diff
@@ -28,16 +28,21 @@
+ * This is a top level Phoenix filter which injected to a scan at the server side. If the scan has
+ * already a filter then PagedFilter wraps it. There two functions this filter implements: paging
+ * and TTL. The paging function makes sure that the scan does not take more than pageSizeInMs.
+ * The TTL function is for masking expired rows.
```
Would it be a cleaner separation of concerns to stack two filters here, `PagedFilter` implements paging, `TTLFilter` implements masking of expired rows?

kadirozde replied:
I could not separate them. Both functionalities needs to be done at the top level filter. If we choose to skip an expired row under or above the PagedFilter then we cannot ensure paging happens timely. I am open to change the name of the class if you have a good suggestion for this.

jpisaac replied:
Isn't server-side paging a configurable feature?

kadirozde replied:
Yes, that is another good reason for separation.

kadirozde replied:
Instead of implementing masking in a filter, I implemented in a region scanner (PhoenixTLLRegionScanner) as the masking logic is more complex than what I thought initially.

apurtell reviewed (approved):
Let me conditionally approve these changes and come back to do a deep dive into the new `StoreCompactionScanner` details. At first glace it looks correct but there are a lot of fiddly internal Phoenix details I am not super familiar with. 

It would also be nice to have more unit test coverage for these changes, specifically, the construction of several predicted edge cases that `StoreCompactionScanner` might encounter. 

If I can use a bit of a tortured analogy, to confirm my understanding of the problem, the issue here is similar to in low level programming what we'd call a torn read or write. On an architecture where the hardware does 32 bit wide manipulations, a long long integer of 64 bits must be updated using multiple instructions. If a thread interrupts another thread after the instruction that does a read or write of the lower or upper 32 bit portion but before the instruction that reads or updates the other portion, and does its own updates, and then thread context switches back, the data has been corrupted, so locking must be introduced to ensure the 64 bit update is atomic. We have the same problem here, except Phoenix is implementing a row addressable relational store, where a row is comprised of one or more cells, over an underlying system that is cell addressable. (HBase does provide a limited fiction of a row based view in some API, notably Filter, but is fundamentally a cell addressable store.) So HBase can expire part of a Phoenix row without preserving Phoenix level required semantics, and the Phoenix row addressable data becomes corrupt ("torn").  

Phoenix must take the design principle that it is solely in charge of data retention. It cannot leave it up to HBase level mechanisms like HBase TTL, cell TTL, and HBase level schema cell scoped retention parameters. It must implement the equivalent functionality in a way that maintains data integrity according to its row addressable requirements. 

There is the new concept of _compaction row version_ introduced in `StoreCompactionScanner` which chronologically sorts mutations and tombstones just like HBase would while also accounting for a row based view. In retrospect this is obvious, which is the hallmark of an elegant solution.

jpisaac commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:
```diff
@@ -426,58 +421,39 @@ RegionScanner getWrappedScanner(final ObserverContext<RegionCoprocessorEnvironme
-        if (isMaxLookbackTimeEnabled(conf)) {
-            setScanOptionsForFlushesAndCompactions(conf, options, store, ScanType.COMPACT_RETAIN_DELETES);
-        }
+        setScanOptionsForFlushesAndCompactions(options);
```
Should the preFlush also have a specialized scanner like the StoreCompactionScanner (or maybe can be reused)?.
As I understand the preFlush currently will keep all rows, versions and delete markers.

kadirozde replied:
Keeping all versions and delete markers is the required behavior for memstore flushes. That is required for indexing consistency and also for max lookback.

jpisaac replied:
But doesn't running the flush thru the StoreCompactionScanner produce the same outcome, i.e keeping the right row versions taking maxLookback and delete markers and all the other settings into account?

kadirozde replied:
But then why to run StoreCompactionScanner if it does not make any difference in this case? The timestamp of these rows in memory will be very recent and they will not expire in practice.

jpisaac commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:
```diff
@@ -34,14 +34,7 @@
-import org.apache.hadoop.hbase.regionserver.ScannerContext;
-import org.apache.hadoop.hbase.regionserver.ScannerContextUtil;
-import org.apache.hadoop.hbase.regionserver.Store;
+import org.apache.hadoop.hbase.regionserver.*;
```
nit: remove import *

kadirozde commented:
> 

@apurtell, Thank you very much for this insightful comment!

jpisaac commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/StoreCompactionScanner.java:
```diff
@@ -0,0 +1,422 @@
+
+    @Override
+    public boolean next(List<Cell> result) throws IOException {
+        synchronized (storeScanner) {
```
@kadirozde Can u add some comments as to why we need the synchronization? Thought we always got a new scanner object for every compact run as in here?
https://github.com/apache/hbase/blob/555bf91f2b5e6240350bc6d279f35165d4596634/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/Compactor.java#L354

kadirozde replied:
I did it for defensive coding. I can remove it.

kadirozde replied:
I did add comments

gjacoby126 commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/StoreCompactionScanner.java:
```diff
@@ -0,0 +1,422 @@
+        scan.readAllVersions();
+        scan.setTimeRange(0, compactionTime);
+        byte[] rowKey = result.get(0).getRowArray();
+        scan.withStartRow(rowKey, true);
```
calling scan.withStartRow twice? Is the second one supposed to be scan.withStopRow(rowKey, true)?

kadirozde replied:
Good catch! Yes, the second one is for the stop row.

gjacoby126 commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/StoreCompactionScanner.java:
```diff
@@ -0,0 +1,422 @@
+        }
+        List<List<Cell>> columns = new ArrayList<>();
+        List<Cell> deleteMarkers = new ArrayList<>();
+        formColumns(result, columns, deleteMarkers);
```
Some high-level comments here would be helpful

gjacoby126 commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/StoreCompactionScanner.java:
```diff
@@ -0,0 +1,422 @@
+                    addCells(rowVersion, result);
+                }
+            } else if (rowVersion.version < maxVersion) {
+                // check if this row needs to be retained
```
no logic here -- the check is elsewhere? If so please mention

kadirozde replied:
I moved the logic for detecting the need for a region level scan to formRowVersions. I will update the code.

gjacoby126 commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/StoreCompactionScanner.java:
```diff
@@ -0,0 +1,422 @@
+    private void updateResultsWithRowOutsideMaxLookbackButInsideTTLWindow(List<Cell> result,
+            CompactionRowVersion rowVersion) {
+        // The rows that are outside the max lookback window but inside the TTL window
+        if (rowVersion.deleteFamilyMarker == null) {
```
Do we need any logic about minVersions in here, or does the fact that minVersions < maxVersions mean that the rowVersion.version < maxVersion check also takes care of minVersion?

kadirozde replied:
The MinVersions check is required to be done outside the TTL window. If the version outside the TTL window is less than MinVersions then it is retained.

gjacoby126 commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/StoreCompactionScanner.java:
```diff
@@ -0,0 +1,422 @@
+        synchronized (storeScanner) {
+            boolean hasMore = storeScanner.next(result);
+            filter(result, true);
+            Collections.sort(result, CellComparator.getInstance());
```
Is there anything we to do about sorting to factor in whether the "new" or "old" Cell ordering is in use? (That is, whether Deletes always win at a particular timestamp)

https://hbase.apache.org/book.html#new.version.behavior

kadirozde replied:
Thank you for the link. I think we need to address this. I will look into it more later.

kadirozde replied:
The secondary indexing implementation let Deletes always win. I preserved the same behavior for Phoenix compaction policy in this PR.

gjacoby126 reviewed (commented):
As Andrew said, a very elegant solution to a hard problem. Some questions and a potential bug:

jpisaac commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/StoreCompactionScanner.java:
```diff
@@ -0,0 +1,422 @@
+                if (firstCell.getType() == Cell.Type.DeleteFamily ||
+                        firstCell.getType() == Cell.Type.DeleteFamilyVersion) {
+                    if (firstCell.getType() == Cell.Type.DeleteFamily) {
+                        if (firstCell.getTimestamp() >= ttlWindowStart &&
```
Don't we want to preserve the delete marker if it is inside the maxLookback window even if KEEP_DELETED_CELLS == FALSE?

kadirozde replied:
Yes, we need to preserve it. The decision is done later in prepareResults. This is just an optimization to early terminate the row version generation here. However, there is a typo in the if statement. It should read if (firstCell.getTimestamp() < maxLookbackWindowStart &&. Good catch!

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/StoreCompactionScanner.java:
```diff
@@ -0,0 +1,422 @@
+                            cell.getQualifierLength(),
+                            dm.getQualifierArray(), dm.getQualifierOffset(),
+                            dm.getQualifierLength()) == 0) {
+                if (dm.getType() == Cell.Type.Delete) {
```
I will add comments here. It is to make sure that a Type.Delete marker is used for deleting only one cell.

kadirozde replied:
Done

stoty commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/UngroupedAggregateRegionObserver.java:
```diff
@@ -588,6 +588,49 @@ public InternalScanner preCompact(ObserverContext<RegionCoprocessorEnvironment>
+                        }
+                    }
+                    if (!isDisabled) {
+                        internalScanner = new StoreCompactionScanner(c.getEnvironment(), store, scanner,
```
IIUC this disables the new scanner for this case, and uses the default one.

TBH I haven't analyzed the new logic in detail, but wouldn't that defeat the original purpose of this patch for this case ?

Also the old code sets the keepDeletedCells and TTL, and this one doesn't. If we use the default scanner, what causes  deletes to be kept ?

stoty replied:
Thanks for the explanation.

stoty commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/UngroupedAggregateRegionObserver.java:
```diff
@@ -588,6 +588,49 @@ public InternalScanner preCompact(ObserverContext<RegionCoprocessorEnvironment>
+                            try (PhoenixConnection conn = QueryUtil.getConnectionOnServer(
+                                    compactionConfig).unwrap(PhoenixConnection.class)) {
+                                PTable table = PhoenixRuntime.getTableNoCache(conn, fullTableName);
+                                List<PTable>
```
We could also check what coprocessors the data table has, and only apply this logic if the old indexing is still used.

kadirozde replied:
Done

kadirozde commented:
I am not done yet. I implemented everything that I planned. I may need to enhance TTL masking to cover one more case where  an expired row revives between compactions. I also need add more comments and possibly more tests. There are still some test failures to address too.

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/StoreCompactionScanner.java:
```diff
@@ -0,0 +1,422 @@
+        compactionTime = EnvironmentEdgeManager.currentTimeMillis();
+        this.maxLookbackWindowStart = compactionTime - maxLookbackInMs;
+        ColumnFamilyDescriptor cfd = store.getColumnFamilyDescriptor();
+        long ttl = cfd.getTimeToLive();
```
I implemented a solution where CompactionScanner maintains a per JVM map of table:column family to max lookback age  so that the max lookback age can be overridden dynamically.

kadirozde commented:
> I am not done yet. I implemented everything that I planned. I may need to enhance TTL masking to cover one more case where an expired row revives between compactions. I also need add more comments and possibly more tests. There are still some test failures to address too.

I have addressed all comments provided so far, fixed test failures and introduced new tests to cover the code. The PR is now finalized. I also added a config param to disable the behavior at the cluster level when needed. @apurtell, @stoty, @gjacoby126, @jpisaac

stoty commented:
The fix for my issue with the old indexing code looks good.
I haven't followed the other issues, so I wouldn't comment on the rest of the patch.

virajjasani commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/UngroupedAggregateRegionObserver.java:
```diff
@@ -588,21 +590,70 @@ public InternalScanner preCompact(ObserverContext<RegionCoprocessorEnvironment>
+                                                + fullTableName, e);
+                            }
+                        }
+                        if (table != null &&
+                                !PhoenixDatabaseMetaData.SYSTEM_CATALOG_NAME.equals(fullTableName) &&
+                                !ScanUtil.hasCoprocessor(c.getEnvironment(),
+                                GlobalIndexChecker.class.getName())) {
+                            List<PTable>
+                                    indexes =
+                                    PTableType.INDEX.equals(table.getType()) ?
+                                            Lists.newArrayList(table) :
+                                            table.getIndexes();
```
Shall we comment just above this if condition that this condition is only for old indexing design?

kadirozde replied:
Yes, I will add it

virajjasani reviewed (approved):
+1, I have ran `PhoenixTableTTLIT` multiple times locally to try covering as much source changes as possible. Overall it looks good, there might be some minor follow ups that can be taken up later but given the criticality of using the new user as well as compaction scanners to mitigate data integrity issues, it's worth merging this PR.

gjacoby126 commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/CompactionScanner.java:
```diff
@@ -0,0 +1,776 @@
+            column.remove(0);
+            if (cell.getType() == Cell.Type.DeleteColumn ||
+                    cell.getType() == Cell.Type.Delete) {
+                rowContext.addColumnDeleteMarker(cell);
```
We add a column delete marker when we hit a row delete marker? (Cell.Type.Delete?)

kadirozde replied:
Cell.Type.Delete is a column delete marker to delete only one cell version. It should have been named Cell.Type.DeleteColumnVersion.

kadirozde commented:
@gjacoby126, I feel that the code is too complicated now and I have just started working on a version that greatly simplifies the compaction logic. I appreciate if you can review it again after that. Hope you did not spend too much time on this version.

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/CompactionScanner.java:
```diff
@@ -0,0 +1,776 @@
+        }
+        Long old = maxLookbackMap.putIfAbsent(tableName + SEPARATOR + columnFamilyName,
+                maxLookbackInMillis);
+        if (old == null || old < maxLookbackInMillis) {
+            maxLookbackMap.put(columnFamilyName, maxLookbackInMillis);
+        }
```
Good catch, it should include the table name

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/PhoenixTTLRegionScanner.java:
```diff
@@ -0,0 +1,218 @@
+                }
+                found = true;
+            }
+            ts = c.getTimestamp();
```
Yes, it is redundant

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/PhoenixTTLRegionScanner.java:
```diff
@@ -0,0 +1,218 @@
+                found = true;
+            }
+            ts = c.getTimestamp();
+            if (maxTimestamp < ts) {
+                maxTimestamp = ts;
+            }
+            if (minTimestamp > ts) {
+                minTimestamp = ts;
+            }
```
Of course they are equal. I did not want to incur a possible function call for every scanned cell unnecessarily. I am not sure if the Java complier will replace these function calls with inline functions. If so, I will happily change it.

kadirozde commented on phoenix-core/src/it/java/org/apache/phoenix/end2end/QueryDatabaseMetaDataIT.java:
```diff
@@ -1126,6 +1126,7 @@ public void testDropAllKVCols() throws Exception {
 
             conn.createStatement().executeUpdate("ALTER TABLE " + tableName + " DROP COLUMN col1");
 
+            TestUtil.dumpTable(conn, TableName.valueOf(tableName));
```
I will remove it

kadirozde commented on phoenix-core/src/it/java/org/apache/phoenix/end2end/SCNIT.java:
```diff
@@ -27,9 +27,11 @@
 import java.sql.SQLException;
 import java.util.Properties;
 
+import org.apache.hadoop.hbase.TableName;
```
Yes

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:
```diff
@@ -21,12 +21,7 @@
-import org.apache.hadoop.hbase.HConstants;
-import org.apache.hadoop.hbase.KeepDeletedCells;
-import org.apache.hadoop.hbase.MemoryCompactionPolicy;
-import org.apache.hadoop.hbase.NotServingRegionException;
```
ok

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:
```diff
@@ -413,59 +435,81 @@ RegionScanner getWrappedScanner(final ObserverContext<RegionCoprocessorEnvironme
+        // We want the store to give us all the deleted cells to StoreCompactionScanner
+        KeepDeletedCells keepDeletedCells = options.getKeepDeletedCells();
+        if (keepDeletedCells == KeepDeletedCells.FALSE) {
+            keepDeletedCells = KeepDeletedCells.TRUE;
```
I will set it to KeepDeletedCells.TTL always regardless of what the existing option is

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/util/ScanUtil.java:
```diff
@@ -1050,6 +1052,17 @@ public static boolean isEmptyColumn(Cell cell, byte[] emptyCF, byte[] emptyCQ) {
 
+
+    public static boolean isEmptyColumn(Cell cell, byte[] emptyCF) {
+        return Bytes.compareTo(cell.getFamilyArray(), cell.getFamilyOffset(),
```
OK

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/CompactionScanner.java:
```diff
@@ -0,0 +1,776 @@
+            }
+        } else {
+            // Deleted
+            if ((rowVersion.version < maxVersion && keepDeletedCells == KeepDeletedCells.TRUE) ||
```
minVersions is considered only outside the TTL window.

kadirozde commented:
@jpisaac, @gjacoby126, @virajjasani, I have completed the code simplification/restructuring and updated the design doc. I also made the changes based on your feedback. You can review the code now. Thank you again!

jpisaac reviewed (commented):
Did an initial review, and have some nit comments. My local testing was also green. Will do a more detailed review of some of the critical classes.

kadirozde commented on phoenix-core/src/it/java/org/apache/phoenix/end2end/ExplainPlanWithStatsEnabledIT.java:
```diff
@@ -165,7 +165,7 @@ public void testBytesRowsForSelectWithLimitIgnored() throws Exception {
             assertFalse(rs.next());
             Estimate info = getByteRowEstimates(conn, sql, binds);
-            assertEquals((Long) 390L, info.estimatedBytes);
+            assertEquals((Long) 691L, info.estimatedBytes);
```
This is to undo my change on 3/17/2019. I had the impression that the Phoenix server code should never return the empty column cell back to the client. It turned out that it was not the case especially for this query. As you can see, this PR does not remove the empty column from the returned result anymore. The increase/revert is because of that.

kadirozde commented on phoenix-core/src/it/java/org/apache/phoenix/end2end/ExplainPlanWithStatsEnabledIT.java:
```diff
@@ -204,7 +204,7 @@ public void testBytesRowsForSelectOnIndex() throws Exception {
             }
             Estimate info = getByteRowEstimates(conn, sql, binds);
-            assertEquals((Long) 390L, info.estimatedBytes);
+            assertEquals((Long) 691L, info.estimatedBytes);
```
See the previous comment.

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/TTLRegionScanner.java:
```diff
@@ -0,0 +1,217 @@
+        ttl = env.getRegion().getTableDescriptor().getColumnFamilies()[0].getTimeToLive();
+        ttlWindowStart = ttl == HConstants.FOREVER ? 1 : currentTime - ttl * 1000;
+        ttl *= 1000;
+        isPhoenixTableTTLEnabled = emptyCF != null && emptyCQ != null &&
```
Regardless if the TTL is disabled cluster wide or the client is an older client and did not supply the empty column parameters, the masking should not be done here. I will rename it to isMaskingEnabled to eliminate any confusion.

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/CompactionScanner.java:
```diff
@@ -0,0 +1,908 @@
+                        nextDeleteFamilyTimestamp = 0;
+                    }
+                }
+            } if (firstCell.getType() == Cell.Type.Put) {
```
In oder to prevent the unnecessary evaluation of the second if condition when the first if condition is true, I wanted to have "else if" but did not type the else keyword apparently. I will add it.

jpisaac commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/CompactionScanner.java:
```diff
@@ -0,0 +1,908 @@
+                    b.getQualifierArray(), b.getQualifierOffset(), a.getQualifierLength());
+        }
+
+        private int compareTypes(Cell a, Cell b) {
```
Maybe a new Comparator?

kadirozde replied:
I simplified trimRegionResult.

jpisaac reviewed (commented):
Added some more comments for CompactionScanner.

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/CompactionScanner.java:
```diff
@@ -0,0 +1,908 @@
+    public CompactionScanner(RegionCoprocessorEnvironment env,
+            Store store,
+            InternalScanner storeScanner,
+            long maxLookbackInMillis,
```
This can be done when a test requiring such a constructor is added. At this moment, it is not required.

jpisaac replied:
I was using a modified constructor for my tests, to enable me to set maxLookbackInMillis, ttl, and compactionTime, hence the ask.

    public CompactionScanner(RegionCoprocessorEnvironment env,
                                Store store,
                                InternalScanner storeScanner,
                                long maxLookbackInMillis,
                                long ttl,
                                long compactionTime,
                                byte[] emptyCF,
                                byte[] emptyCQ) {

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/CompactionScanner.java:
```diff
@@ -0,0 +1,908 @@
+     * Any coprocessors within a JVM can extend the max lookback window for a column family
+     * by calling this static method.
+     */
+    public static void overrideMaxLookback(String tableName, String columnFamilyName,
```
I looked at sharedData but did not see how it would improve the implementation. We are not really sharing data here. Actually, what is implemented here provides better data hiding.

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/CompactionScanner.java:
```diff
@@ -0,0 +1,908 @@
+         * KeepDeletedCells is TRUE
+         *
+         */
+        private boolean retainInsideTTLWindow(List<Cell> result,
```
Good point! We do not need these methods to return a value anymore after restructuring the code.

jpisaac commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:
```diff
@@ -354,10 +361,14 @@ public boolean nextRaw(List<Cell> result) throws IOException {
+            @Override
+            public RegionScanner getNewRegionScanner(Scan scan) throws IOException {
+                return new RegionScannerHolder(c, scan,
+                        ((DelegateRegionScanner) delegate).getNewRegionScanner(scan));
```
Coercing without checking the instance is of type DelegateRegionScanner can lead to ClassCastException.

kadirozde replied:
That is the expected behavior.

kadirozde replied:
Added try-catch for ClassCastException

jpisaac reviewed (commented):
Reviewed TTLRegionScanner

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/TTLRegionScanner.java:
```diff
@@ -0,0 +1,220 @@
+        this.scan = scan;
+        emptyCQ = scan.getAttribute(EMPTY_COLUMN_QUALIFIER_NAME);
+        emptyCF = scan.getAttribute(EMPTY_COLUMN_FAMILY_NAME);
+        long currentTime = EnvironmentEdgeManager.currentTimeMillis();
```
Absolutely. I actually planned to make this change but forgot it. Thank you for catching this!

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/TTLRegionScanner.java:
```diff
@@ -0,0 +1,220 @@
+    private void init() throws IOException {
+        // HBase PageFilter will also count the expired rows.
+        // Instead of using PageFilter for counting, we will count returned row here.
+        PageFilter pageFilter = ScanUtil.removePageFilter(scan);
```
PageFilter is an HBase builtin filter. The name should have been LimitFilter instead of PageFilter. So, there is no relation between the server paging feature and PageFilter. Phoenix has PagingRegionScanner and PagingFilter. I know it is confusing. 

The code is simpler when we count always.  Adding an if statement and counting only when the PageFilter is removed is not more efficient.

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/TTLRegionScanner.java:
```diff
@@ -0,0 +1,220 @@
+            }
+            Cell cell = result.get(0);
+            result.clear();
+            if (EnvironmentEdgeManager.currentTimeMillis() - startTime > pageSize) {
```
Good catch! This one was supposed to be pageSizeInMs. I will fix this.

kadirozde commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/DelegateRegionScanner.java:
```diff
@@ -87,4 +88,8 @@ public int getBatch() {
     }
+
+    public RegionScanner getNewRegionScanner(Scan scan) throws IOException {
+        return ((DelegateRegionScanner)delegate).getNewRegionScanner(scan);
```
Regardless if you check or not, when the instance is not of the expected type, an exception should be thrown.

kadirozde replied:
Added try-catch for ClassCastException

jpisaac commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/CompactionScanner.java:
```diff
@@ -0,0 +1,879 @@
+         * The generates the intersection of regionResult and input. The result is the resulting
+         * intersection.
+         */
+        private void trimRegionResult(List<Cell> regionResult, List<Cell> input,
```
Don't quite understand this, will sync with you offline

jpisaac commented:
@kadirozde Other than some clarifications I am done with my review. Let me know if you have any more updates

jpisaac reviewed (approved):
LGMT +1, Thank you @kadirozde for the fix and enhancements.
